python/ray/util/multiprocessing/pool.py

```diff
@@ -192,40 +194,52 @@ def add_object_ref(self, object_ref):
     def run(self):
         unready = copy.copy(self._object_refs)
         aggregated_batch_results = []
-        while self._num_ready < self._total_object_refs:
+        # Run for a specific number of objects if self._total_object_refs > 0
+        # Run indefinitely if self._total_object_refs == 0, until stop event
+        while (self._total_object_refs == 0) or (
```

Comment:
Could we refactor this loop a bit? Could we change this top `while` condition to be:

```python
while (self._num_ready < self._total_object_refs) or (self._total_object_refs == 0 and not self._stop_event.is_set()):
    ...
```

Then, could we replace [the if-else condition in line 217](https://github.com/ray-project/ray/pull/24117/files#diff-887a588c8b872641a8214860aa7d8375aa2948669c9c8d39518618939fcd2f4cR217):

```python
if len(unready) == 0:
    # not awaiting any IDs; stop if the stop event has been received
    if self._stop_event.is_set():
        break
else:
    [ready_id], unready = ray.wait(unready, num_returns=1)
    ...
```

with an if-condition and a comment:

```python
# This condition may be False if self._stop_event is set
# and no batch elements are processed in this iteration.
if len(unready) > 0:
    [ready_id], unready = ray.wait(unready, num_returns=1)
    ...
```

This removes the body of the current if-clause, and it makes the code a bit more readable in my opinion. Do you see any concurrency (or other) issues with this?

Reply:
The problem with the while condition is that it would end the while loop once _stop_event  becomes true. Because it only processes one result per loop (`ray.wait(unready, num_returns=1)`), that would mean we may terminate before all results have been collected.

For this reason we need an explicit break once both `unready` is empty *and* `_stop_event.is_set()`, I think.

Reply:
I see– I wonder if there are other ways to simplify this. Let me confirm my understanding–

* When `_stop_event` is set, a `None` value is added to `self._object_refs`.
* Currently, `_stop_event` gets set only once all the elements in the iterable are consumed, and `next()` is called at least one more time.
* Since `next()` processes each element (or times out) before it is called again, wouldn't it be impossible for `_stop_event` to get set until all the results get collected?

Reply:
Actually, what's the reason we use `_stop_event`? Shouldn't the `None` value be enough to indicate when the thread should stop?

Reply:
Your comments made me reconsider the structure I was using. I am now pushing a version with quite a few changes. 

`IMapIterator._submit_next_chunk` now indicates when the final chunk is submitted, using `self._result_thread.add_object_ref(new_chunk_id, final=self._finished_iterating)`.

This is used in the `add_object_ref` function to push an additional `None` object. This in turn set the `winding_down` flag in the `run()` thread, so that the run function exits on its own. This avoids the need to explicitly shut down the `run()` thread on the n+1 iteration.

Reply:
I do wonder whether it's useful to make the two modes of operation for `ResultThread` even more explicit: 
* In 'fixed length' mode, it would operate as before. Adding 'final' objects would be invalid, and `None` objects would be ignored.
* In 'variable length' mode, it would wait for the addition of a new object with a final flag (although it would also end if a `None` object is manually pushed - this could be prevented with an argument check.

As `ResultThread` is only created internally, do you think it's necessary to implement the necessary checks, just to avoid future (accidental) misuse?

Reply:
Nice work on the new structure! I like the improvements. I'm wondering if there's a way we can unify the two paths, so we don't have to maintain them separately:

* Make `total_object_refs=float('inf')` when we're using an iterable of unknown length. Then we can make the [outer while-loop condition](https://github.com/ray-project/ray/blob/abfd82306cdbe625c47c582578538725ffcb8879/python/ray/util/multiprocessing/pool.py#L203-L205) in `run`:

```python
while self._num_ready < self._total_object_refs:
```

* We can remove all the logic with the `winding_down` variable. When we receive the [`END_SENTINEL`](https://github.com/ray-project/ray/pull/24117/files#r857976076) in `run`, we can set `self._total_object_refs = len(self._object_refs)` instead of `winding_down=True`:

```python
if new_object_ref is END_SENTINEL:
    self._total_object_refs = len(self._object_refs)
```

Then, I believe the existing path for `self._total_object_refs` should handle the rest. What do you think?

Reply:
This is a brilliant suggestion! I have implemented it and it works flawlessly, with much cleaner code.